pkg/services/hcloud/server/server.go

```diff
@@ -191,6 +248,411 @@ func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, e
+		hm.SetBootState(infrav1.HCloudBootStateWaitForRebootAfterImageURLCommandThenBootToRealOS)
+		return reconcile.Result{RequeueAfter: requeueImmediately}, nil
+
+	case sshclient.ImageURLCommandStateFinishedFailed:
+		msg := "ImageURLCommand failed. Deleting machine"
+		err = errors.New(msg)
+		s.scope.Logger.Error(err, "",
+			"logFile", logFile)
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return reconcile.Result{}, nil
+	case sshclient.ImageURLCommandStateNotStarted:
+		return reconcile.Result{}, fmt.Errorf("image-url-command not started in BootState %q? Should not happen",
+			state)
+	default:
+		return reconcile.Result{}, fmt.Errorf("unknown ImageURLCommandState: %q", state)
+	}
+}
+
+func (s *Service) handleBootStateWaitForRebootAfterImageURLCommandThenBootToRealOS(ctx context.Context, server *hcloud.Server) (res reconcile.Result, err error) {
+	hm := s.scope.HCloudMachine
+	updateHCloudMachineStatusFromServer(hm, server)
+
+	if hm.Status.ActionIDRebootAfterImageURLCommand == 0 {
+		msg := "handleBootStateWaitForRebootAfterImageURLCommandThenBootToRealOS ActionIDRebootAfterImageURLCommand not set? Can not continue. Provisioning Failed"
+		s.scope.Logger.Error(nil, msg)
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return reconcile.Result{}, nil
+	}
+
+	if hm.Status.ActionIDRebootAfterImageURLCommand != actionDone {
+		action, err := s.scope.HCloudClient.GetAction(ctx, hm.Status.ActionIDRebootAfterImageURLCommand)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("GetAction failed: %w", err)
+		}
+
+		if action.Finished.IsZero() {
+			// not finished yet.
+			conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+				string(hm.Status.BootState), clusterv1.ConditionSeverityInfo,
+				"Waiting until Action RebootAfterImageURLCommand is finished")
+			return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		err = action.Error()
+		if err != nil {
+			err = fmt.Errorf("action %+v failed (Reboot after hcloud-image-url-command): %w", action, err)
+			s.scope.Logger.Error(err, "")
+			msg := err.Error()
+			s.scope.SetError(msg, capierrors.CreateMachineError)
+			conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+				string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+				"%s", msg)
+			return reconcile.Result{}, nil
+		}
+
+		s.scope.Logger.Info("Action RebootAfterImageURLCommand is finished",
+			"actionDuration", action.Finished.Sub(action.Started),
+			"finishedSince", time.Since(action.Finished),
+			"actionStatus", action.Status)
+
+		hm.Status.ActionIDRebootAfterImageURLCommand = actionDone
+	}
+	hm.SetBootState(infrav1.HCloudBootStateBootToRealOS)
```

Comment:
why exactly do you need to go into this state? This next state is exactly what you already do in the very first state

Reply:
in the old code I waited for the action to be finished. But this is gone, one BootState less :-)